### PR TITLE
Store L2 > L1 messages in storage and return them in transaction receipt

### PR DESCRIPTION
**Pull request by m-kus**
In order to aggregate Starknet state (includes block header fields and messages) we need to keep L2 > L1 messages in the storage (L1 > L2 messages can be restored from L1Handler operations).

Related to #1133

### CODE REVIEW

File: crates/pallets/starknet/src/tests/send_message.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+// Generate: starkli class-hash ./cairo-contracts/build/send_message.json
+const SEND_MESSAGE_CLASS_HASH: &str = "0x024e12108da2b3d3c80c7bc99aacdec97e445199c5a8ded7435bf77ce4507631";
+// const SEND_MESSAGE_TO_L2_SELECTOR: &str =
+// "0x38fb0bf48fe489ae23d0a1d7f2b7195ec0b94bfeb2f408b13bfd943d8410d72";
+const SEND_MESSAGE_TO_L1_SELECTOR: &str = "0x9139dbd19ca9654d773cd88f31af4c8d583beecc3362fb986dccfef5cf134f";
+
+// Troubleshooting:
+// Add println! to the necessary runtime method (e.g. print CallInfo structs or extended blockifier
+// message) RUST_LOG=runtime=error RUST_BACKTRACE=1 cargo test --package pallet-starknet
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/send_message.rs:23]: Why is `RUST_LOG=runtime=error` needed if you debug through `println`?

File: crates/pallets/starknet/src/tests/send_message.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            get_account_address(None, AccountType::V0(AccountTypeV0Inner::NoValidate)).into();
+        let contract_class = get_contract_class("send_message.json", 0);
+        let class_hash = Felt252Wrapper::from_hex_be(SEND_MESSAGE_CLASS_HASH).unwrap();
+
+        let declare_tx = DeclareTransactionV1 {
+            sender_address,
+            class_hash,
+            nonce: Felt252Wrapper::ZERO,
+            max_fee: u128::MAX,
+            signature: bounded_vec!(),
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/send_message.rs:41]: Is is not supposed to be bounded anymore?
Did we miss one in the codebase?

Comment (tdelabro) [reply crates/pallets/starknet/src/tests/send_message.rs:41]: Same for all other bounded_vec in the pallet

File: crates/pallets/starknet/src/tests/send_message.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        };
+
+        assert_ok!(Starknet::invoke(RuntimeOrigin::none(), invoke_tx.clone().into()));
+
+        let chain_id = Starknet::chain_id();
+        let tx_hash = invoke_tx.compute_hash::<<MockRuntime as Config>::SystemHash>(chain_id, false);
+        let messages = Starknet::tx_messages(TransactionHash::from(tx_hash));
+
+        assert_eq!(1, messages.len());
```
Comment (tdelabro) [inline crates/pallets/starknet/src/tests/send_message.rs:85]: Can you also check the message content?

File: starknet-rpc-test/get_transaction_receipt.rs [hunk 4]
```diff
@@ -265,3 +268,91 @@ async fn fail_invalid_transaction_hash(#[future] madara: MadaraClient) -> Result
...
+
+    let invoke_tx_receipt = get_transaction_receipt(rpc, invoke_tx_hash).await?;
+
+    match invoke_tx_receipt {
+        MaybePendingTransactionReceipt::Receipt(TransactionReceipt::Invoke(receipt)) => {
+            assert_eq_msg_to_l1(
+                vec![MsgToL1 {
+                    from_address: contract_address,
+                    to_address: FieldElement::ONE, // hardcoded in the contract
+                    payload: vec![],               // no payload
```
Comment (tdelabro) [inline starknet-rpc-test/get_transaction_receipt.rs:349]: Can you try with a payload?

File: starknet-rpc-test/get_transaction_receipt.rs [hunk 5]
```diff
@@ -265,3 +268,91 @@ async fn fail_invalid_transaction_hash(#[future] madara: MadaraClient) -> Result
...
+    // 2. Determine class hash
+
+    let class_hash = match txs.remove(0).unwrap() {
+        TransactionResult::Declaration(rpc_response) => rpc_response.class_hash,
+        _ => panic!("expected execution result"),
+    };
+
+    // 3. Next, deploying an instance of this class using universal deployer
+
+    let deploy_tx = account.call_contract(
```
Comment (tdelabro) [inline starknet-rpc-test/get_transaction_receipt.rs:294]: In starknet `calls` are not supposed to modify the state. 
Therefore it is not supposed to actually deploy the contract. Just to do a `dry` run

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: cairo-contracts/src/test_contracts/send_message.cairo [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+@external
+func send_message_l2_to_l1{syscall_ptr : felt*}() {
+    let (payload: felt*) = alloc();
+    send_message_to_l1(1, 0, payload);
+    return ();
+}
+
+@l1_handler
+func send_message_l1_to_l2(from_address: felt) {
+    assert from_address = 1;
```
[the remark this answers is not in the document]
Comment (m-kus) [reply cairo-contracts/src/test_contracts/send_message.cairo:15]: Yes, I wanted to use it later for testing both types of messages, will update according to the validation rules we'll have for l1 handler txs

File: crates/pallets/starknet/src/lib.rs [hunk 7]
```diff
@@ -1033,6 +1058,53 @@ impl<T: Config> Pallet<T> {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `call_info` — A ref to the call info structure.
+    /// * `next_order` — Next expected message order, has to be 0 for a top level invocation
+    ///
+    /// # Returns
+    ///
+    /// Next expected message order
+    fn aggregate_messages_in_call_info(tx_hash: TransactionHash, call_info: &CallInfo, next_order: usize) -> usize {
```
[the remark this answers is not in the document]
Comment (m-kus) [reply crates/pallets/starknet/src/lib.rs:1071]: I though about it, probably via visitor patter, will try to refactor

File: crates/pallets/starknet/src/tests/send_message.rs (hunk 1 above)
Comment (m-kus) [reply crates/pallets/starknet/src/tests/send_message.rs:23]: Good question :)
I need to update the comment, it's just the notes for myself in order to to forget

File: crates/pallets/starknet/src/tests/send_message.rs (hunk 2 above)
Comment (m-kus) [reply crates/pallets/starknet/src/tests/send_message.rs:41]: Ah, I think I copied that from the old tests before refacto. Will change to vec![]

File: starknet-rpc-test/get_transaction_receipt.rs (hunk 5 above)
Comment (m-kus) [reply starknet-rpc-test/get_transaction_receipt.rs:294]: Ah ok, it's just the matter of naming or other changes required?

File: crates/pallets/starknet/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply crates/pallets/starknet/src/lib.rs:1071]: It's not absolutely crucial. If it add too may traits, abstractions or overhead, we can keep a lil bit of code duplication :)

[the remark this answers is not in the document]
Comment (m-kus) [reply crates/pallets/starknet/src/lib.rs:1047]: So, we can dedup some code, but it becomes much less readable, it's probably better to keep it as is

File: starknet-rpc-test/get_transaction_receipt.rs (hunk 5 above)
Comment (tdelabro) [reply starknet-rpc-test/get_transaction_receipt.rs]: If you are doing an invoke and not a call, yes just rename it will work

**Review by tdelabro — APPROVED** [review]

**Pull request merged**